ListUtil.xs

```diff
@@ -1217,12 +1221,33 @@ CODE:
 #endif
             }
 
-            if(!SvOK(arg) || SvUOK(arg))
+            if(!SvOK(arg) || SvUOK(arg)) {
                 sv_setpvf(keysv, "%" UVuf, SvUV(arg));
-            else if(SvIOK(arg))
+            }
+            else if(SvIOK(arg)) {
                 sv_setpvf(keysv, "%" IVdf, SvIV(arg));
-            else
-                sv_setpvf(keysv, "%.15" NVgf, SvNV(arg));
+            }
+            else {
+                nv_arg = SvNV(arg);
+                /* use 0 for both 0 and -0.0 */
+                if(nv_arg == 0) {
+                    sv_setpvs(keysv, "0");
+                }
+                /* for NaN, use the platform's normal stringification */
+                else if (nv_arg != nv_arg) {
+                    sv_setpvf(keysv, "%" NVgf, nv_arg);
+                }
+                /* for numbers outside of the IV or UV range, we don't need to
+                 * use a comparable format, so just use the raw bytes */
+                else if (nv_arg < (NV)IV_MIN || nv_arg > (NV)UV_MAX) {
+                    sv_setpvn(keysv, (char *) &nv_arg, sizeof(NV));
+                }
+                /* smaller floats get formatted using %g and could be equal to
+                 * a UV or IV */
+                else {
+                    sv_setpvf(keysv, "%0." STRINGIZE(NV_MAX_PRECISION) NVgf, nv_arg);
```

Comment:
Rather than this `STRINGIZE` macro, you can just use the "%0.*g" ability of `sprintf`:

```
    sv_setpvf(keysv, "%0.*" NVgf, NV_MAX_PRECISION, nv_arg);
```

Changing the code to this keeps all the tests passing for me; you can then delete the macro at the top.